open_spiel/games/dark_hex.cc

```diff
@@ -0,0 +1,267 @@
+  /*min_num_players=*/2,
+  /*provides_information_state_string=*/true,
+  /*provides_information_state_tensor=*/true,
+  /*provides_observation_string=*/true,
+  /*provides_observation_tensor=*/true,
+  /*parameter_specification=*/
+  {
+    {"obstype", GameParameter(static_cast<std::string>(kDefaultObsType))},
+    {"board_size", GameParameter(hex::kDefaultBoardSize)},
+  }
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params){
+  return std::shared_ptr<const Game>(
+    new DarkHexGame(params)
+  );
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+} // namespace
+
+DarkHexState::DarkHexState(std::shared_ptr<const Game> game, int board_size,
+                                  ObservationType obs_type)
+    : State(game), state_(game, board_size), board_size_(board_size), 
+                          obs_type_(obs_type) {
+  black_view_.resize(board_size * board_size, CellState::kEmpty);
+  white_view_.resize(board_size * board_size, CellState::kEmpty);
+
+  kNumCells = board_size_ * board_size_;
+  kBitsPerAction = kNumCells + 1;
+  kLongestSequence = kNumCells * 2 - 1;
+}
+
+void DarkHexState::DoApplyAction(Action move) {
+  Player cur_player = CurrentPlayer(); //current player
+  auto& cur_view = (cur_player == Player(0) ? black_view_ : white_view_);
+
+  // Either occupied or not
+  if (state_.BoardAt(move) == CellState::kEmpty) {
+    state_.ApplyAction(move);
+  } else {}
+
+  SPIEL_CHECK_EQ(cur_view[move], CellState::kEmpty);
+  cur_view[move] = state_.BoardAt(move);
+  action_sequence_.push_back(std::pair<int,Action>(cur_player, move));
+}
+
+std::vector<Action> DarkHexState::LegalActions() const {
+  if (IsTerminal()) return {};
+  std::vector<Action> moves;
+  const Player player = CurrentPlayer();
+  const auto& cur_view = (player == Player{0} ? black_view_ : white_view_);
+
+  for (Action move = 0; move < kNumCells; ++move){
+    if (cur_view[move] == CellState::kEmpty){
+      moves.push_back(move);
+    }
+  }
+
+  return moves;
+}
+
+std::string DarkHexState::ViewToString(Player player) const {
+  const auto& cur_view = (player == Player{0} ? black_view_ : white_view_);
+  std::string str;
+
+  // TODO: ??
+  // Change here for row & cols after customizing for x, y
+  int num_rows = board_size_, num_cols = board_size_;
+  for (int r = 0; r < num_rows; ++r){
+    for (int c = 0; c < num_cols; ++c){
+      absl::StrAppend(&str, StateToString(cur_view[r * num_cols + c]));
+    }
+    if (r < (num_rows - 1)) absl::StrAppend(&str, "\n");
+  }
+  return str;
+}
+
+std::string DarkHexState::ActionSequenceToString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  std::string str;
+  for (const auto& player_with_action: action_sequence_) {
+    if (player_with_action.first == player) {
+      str.append(std::to_string(player_with_action.first));
+      str.push_back(',');
+      str.append(std::to_string(player_with_action.second));
+      str.push_back(' ');
```

Comment:
Can you use absl::StrAppend(&str, ...)? See the other games for examples.

Reply:
Sure will do